cpp/src/copying/contiguous_split.cu

```diff
@@ -14,466 +14,1105 @@
+  cur_h_buf += indices_size;
+  src_buf_info* h_src_buf_info = reinterpret_cast<src_buf_info*>(cur_h_buf);
+  cur_h_buf += src_buf_info_size;
+  size_t* h_buf_sizes = reinterpret_cast<size_t*>(cur_h_buf);
+  cur_h_buf += buf_sizes_size;
+  dst_buf_info* h_dst_buf_info = reinterpret_cast<dst_buf_info*>(cur_h_buf);
+  cur_h_buf += dst_buf_info_size;
+  uint8_t** h_src_bufs = reinterpret_cast<uint8_t**>(cur_h_buf);
+  cur_h_buf += src_bufs_size;
+  uint8_t** h_dst_bufs = reinterpret_cast<uint8_t**>(cur_h_buf);
+
+  // compute stack space needed for nested list offset calculation (needed on gpu only)
+  int offset_stack_partition_size = compute_offset_stack_size(input.begin(), input.end());
+  int offset_stack_size           = offset_stack_partition_size * num_partitions * 4;
+
+  // allocate device
+  size_t total_device_temp_size = total_temp_size + offset_stack_size;
+  rmm::device_buffer device_buf{total_device_temp_size, stream, mr};
+  // distribute
+  uint8_t* cur_d_buf   = reinterpret_cast<uint8_t*>(device_buf.data());
+  size_type* d_indices = reinterpret_cast<size_type*>(cur_d_buf);
+  cur_d_buf += indices_size;
+  src_buf_info* d_src_buf_info = reinterpret_cast<src_buf_info*>(cur_d_buf);
+  cur_d_buf += src_buf_info_size;
+  size_t* d_buf_sizes = reinterpret_cast<size_t*>(cur_d_buf);
+  cur_d_buf += buf_sizes_size;
+  dst_buf_info* d_dst_buf_info = reinterpret_cast<dst_buf_info*>(cur_d_buf);
+  cur_d_buf += dst_buf_info_size;
+  uint8_t** d_src_bufs = reinterpret_cast<uint8_t**>(cur_d_buf);
+  cur_d_buf += src_bufs_size;
+  uint8_t** d_dst_bufs = reinterpret_cast<uint8_t**>(cur_d_buf);
+  cur_d_buf += dst_bufs_size;
+  size_type* d_offset_stack = reinterpret_cast<size_type*>(cur_d_buf);
+
+  // compute splits -> indices.
+  {
+    size_type* indices = h_indices;
+    *indices           = 0;
+    indices++;
+    std::for_each(splits.begin(), splits.end(), [&indices](auto split) {
+      *indices = split;
+      indices++;
+    });
+    *indices = input.column(0).size();
+
+    for (size_t i = 0; i < splits.size(); i++) {
+      auto begin = h_indices[i];
+      auto end   = h_indices[i + 1];
+      CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+      CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+      CUDF_EXPECTS(end <= input.column(0).size(), "Slice range out of bounds.");
+    }
+  }
+
+  // setup source buf info
+  setup_source_buf_info(input.begin(), input.end(), h_src_buf_info, {h_src_buf_info, 0});
+
+  // HtoD indices and source buf info to device
+  cudaMemcpyAsync(
+    d_indices, h_indices, indices_size + src_buf_info_size, cudaMemcpyHostToDevice, stream);
+  cudaStreamSynchronize(stream);
+
+  // compute sizes of each column in each partition, including alignment.
+  thrust::transform(
+    rmm::exec_policy(stream)->on(stream),
+    thrust::make_counting_iterator<size_t>(0),
+    thrust::make_counting_iterator<size_t>(num_bufs),
+    d_dst_buf_info,
+    [num_src_bufs,
```

Comment:
This feels like a raw loop masquerading as an algorithm.

Reply:
How so?  It's producing a processed dst_info_buf per input.

Reply:
I'm not sure what the alternative is here.  This is happening on the gpu, so the other options are:
- Use a thrust::for_each which will have the same issues
- Write a kernel which will be similar because it has to take all these parameters